TombEngine/Objects/TR4/Trap/SquishyBlock.cpp

```diff
@@ -0,0 +1,285 @@
+			}
+			else if (pointColl.FloorTilt.x < 0)
+			{
+				slopeAngle = ANGLE(90.0f);
+			}
+
+			if (pointColl.FloorTilt.y > 0 && pointColl.FloorTilt.y > abs(pointColl.FloorTilt.x))
+			{
+				slopeAngle = ANGLE(180.0f);
+			}
+			else if (pointColl.FloorTilt.y < 0 && -pointColl.FloorTilt.y > abs(pointColl.FloorTilt.x))
+			{
+				slopeAngle = ANGLE(0.0f);
+			}
+
+			short dirAngle = phd_atan(dir.z, dir.x);
+			short alignAngle = Geometry::GetShortestAngle(slopeAngle, dirAngle);
+
+			// Test if slope aspect is aligned with direction.
+			if (alignAngle != 0 && alignAngle != ANGLE(180.0f))
+				return false;
+		}
+
+		// Check for diagonal split.
+		if (pointColl.Position.DiagonalStep)
+			return false;
+
+		// Test ceiling height.
+		int relCeilHeight = abs(pointColl.Position.Ceiling - pointColl.Position.Floor);
+		
+		if (relCeilHeight <= itemHeight)
+			return false;
+
+		// Check for blocked grey box.
+		if (g_Level.Boxes[pointColl.Block->Box].flags & BLOCKABLE)
+		{
+			if (g_Level.Boxes[pointColl.Block->Box].flags & BLOCKED)
+				return false;
+		}
+
+		// Check for inaccessible sector.
+		if (pointColl.Block->Box == NO_ZONE)
+			return false;
+
+		// Check for stopper flag.
+		if (pointColl.Block->Stopper)
+			return false;
+
+		return true;
+	}
+
+	void SquishyBlockCollision(short itemNumber, ItemInfo* laraItem, CollisionInfo* coll)
+	{
+		auto& item = g_Level.Items[itemNumber];
+
+		if (TestBoundsCollide(&item, laraItem, coll->Setup.Radius) && TestCollision(&item, laraItem))
+		{
+			if (laraItem->HitPoints > 0)
+				ItemPushItem(&item, laraItem, coll, false, 1);			
+		}
+
+		if (ItemPushItem(&item, laraItem, coll, false, 1))
+		{
+			if (item.Animation.ActiveState == SQUISHY_BLOCK_STATE_ORIGINAL)
+			{
+				auto frame = item.Animation.FrameNumber - GetAnimData(item).frameBase;
+				if (!frame || frame == 33)
+					LaraItem->HitPoints = 0;
+			}
+			else if (item.Animation.ActiveState == SQUISHY_BLOCK_STATE_COLLIDE_RIGHT ||
+					item.Animation.ActiveState == SQUISHY_BLOCK_STATE_COLLIDE_LEFT)
+			{
+				LaraItem->HitPoints = 0;
+			}
+		}	
+	}
+
+	void FallingSquishyBlockCollision(short itemNumber, ItemInfo* laraItem, CollisionInfo* coll)
+	{
+		auto& item = g_Level.Items[itemNumber];
+
+		if (TestBoundsCollide(&item, laraItem, coll->Setup.Radius) && TestCollision(&item, laraItem))
+		{
+			if (item.Animation.FrameNumber - GetAnimData(item).frameBase <= 8)
+			{
+				item.Animation.FrameNumber += 2;
+				laraItem->HitPoints = 0;
+				SetAnimation(laraItem, LA_BOULDER_DEATH);
+				laraItem->Animation.Velocity.z = 0;
+				laraItem->Animation.Velocity.y = 0;		
+			}
+			else if (laraItem->HitPoints > 0)
+				ItemPushItem(&item, laraItem, coll, false, 1);
+		}
+	}
+
+	void ControlFallingSquishyBlock(short itemNumber)
+	{
+		auto& item = g_Level.Items[itemNumber];
+
+		if (TriggerActive(&item))
+		{
+			if (item.ItemFlags[0] < 60)
+			{
+				SoundEffect(SFX_TR4_EARTHQUAKE_LOOP, &item.Pose);
+				Camera.bounce = (item.ItemFlags[0] - 92) >> 1;
```

Comment:
Can we avoid using bitshift here? Meaning of this operation is not clear in this context.

Reply:
fixed